inductor: make onednn linear inputs are always real contiguous

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108635
* __->__ #108560


For OneDNN linear, if packed linear inputs are not the default contiguous tensor, it always calls in ref pat and gets a worse performance, this PR will force its inputs to the actual default contiguous tensor.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

XiaobingSuper committed:
inductor: make onednn fuse linear inputs are always real contiguous

[ghstack-poisoned]

jgong5 reviewed (commented):
Is it a regression due to onednn upgrade? Perhaps better to address in onednn instead?

jgong5 reviewed (approved):
Realized that we require input as contiguous in the c++ kernel so doing `require_contiguous` in the inductor should be correct.

desertfire commented on test/inductor/test_cpu_repro.py:
```diff
@@ -2466,6 +2466,28 @@ def fn(x, y):
+    def test_linear_with_no_default_contiguous_input(self):
+        mod = torch.nn.Sequential(torch.nn.Linear(16, 16)).eval()
+        temp = torch.randn(1, 16, 1, 1)
+        v = torch.ops.inductor._reinterpret_tensor(temp, [1, 16], [0, 1], 0)
```
Nit: why using `_reinterpret_tensor` instead of `as_strided`? `_reinterpret_tensor` is meant to be an internal op.

XiaobingSuper replied:
Updated, using as_strided now.

desertfire reviewed (approved):
If you will address my comment to `test_linear_with_no_default_contiguous_input`, you may also consider fixing the PR title (as in grammar).

XiaobingSuper committed:
Update on "inductor: make onednn linear inputs are always real contiguous"



For OneDNN linear, if packed linear inputs are not the default contiguous tensor, it always calls in ref pat and gets a worse performance, this PR will force its inputs to the actual default contiguous tensor.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: make onednn linear inputs are always real contiguous"



For OneDNN linear, if packed linear inputs are not the default contiguous tensor, it always calls in ref pat and gets a worse performance, this PR will force its inputs to the actual default contiguous tensor.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge

XiaobingSuper commented:
@pytorchbot merge
